src/utils.ts

```diff
@@ -0,0 +1,564 @@
+
+  const data = response.data;
+
+  if (!data.success) {
+    throw new Error(`Block height ${height} download failed`);
+  }
+
+  const responseBlock = data.block;
+
+  const block: FullBlock = {
+    txId: responseBlock.tx_id as string,
+    version: responseBlock.version as number,
+    weight: responseBlock.weight as number,
+    timestamp: responseBlock.timestamp as number,
+    isVoided: responseBlock.is_voided as boolean,
+    inputs: responseBlock.inputs.map((input) => {
+      const typedDecodedScript: DecodedScript = {
+        type: input.decoded.type as string,
+        address: input.decoded.address as string,
+        timelock: input.decoded.timelock ? input.decoded.timelock as number : null,
+        value: input.decoded.value ? input.decoded.value as number : null,
+        tokenData: input.decoded.token_data ? input.decoded.token_data as number : null,
+      };
+      const typedInput: Input = {
+        txId: input.tx_id as string,
+        index: input.index as number,
+        value: input.value as number,
+        tokenData: input.token_data as number,
+        script: input.script as string,
+        decoded: typedDecodedScript,
+        token: input.token as string,
+      };
+
+      return typedInput;
+    }),
+    outputs: responseBlock.outputs.map((output): Output => {
+      const typedDecodedScript: DecodedScript = {
+        type: output.decoded.type as string,
+        address: output.decoded.address as string,
+        timelock: output.decoded.timelock ? output.decoded.timelock as number : null,
+        value: output.decoded.value ? output.decoded.value as number : null,
+        tokenData: output.decoded.token_data ? output.decoded.token_data as number : null,
+      };
+
+      const typedOutput: Output = {
+        value: output.value as number,
+        tokenData: output.token_data as number,
+        script: output.script as string,
+        decoded: typedDecodedScript,
+        token: output.token as string,
+        spentBy: output.spent_by ? output.spent_by as string : null,
+      };
+
+      return typedOutput;
+    }),
+    parents: responseBlock.parents,
+    height: responseBlock.height as number,
+  };
+
+  return block;
+};
+
+/**
+ * Recursively downloads all transactions that were confirmed by a given block
+ *
+ * @param blockId - The blockId to download the transactions
+ * @param txIds - List of transactions to download
+ * @param data - Downloaded transactions, used while being called recursively
+ */
+export const recursivelyDownloadTx = async (blockId: string, txIds: string[] = [], data: FullTx[] = []): Promise<FullTx[]> => {
+  if (txIds.length === 0) {
+    return data;
+  }
+
+  const txId = txIds.pop();
+  const txData = await downloadTx(txId);
+  const { tx, meta } = txData;
+  const parsedTx: FullTx = parseTx(tx);
+
+  if (parsedTx.parents.length > 2) {
+    // We downloaded a block, we should ignore it
+    return recursivelyDownloadTx(blockId, txIds, data);
```

Comment:
Why do we call it again? I feel we could be diving too deep into the recursion using this strategy. I was expecting to simply return when we want to skip something.

Maybe we should have an outside loop calling this method for all `txIds`.

Reply:
I was thinking about an "impossible" situation where the fullnode API could answer with a block on `parents[1]` or `parents[2]` for some reason

In this case, we want to continue download the transaction list on `txIds`

By default, we never call this method with blocks, so we would never dive too deep

Reply:
That's impossible by design. Blocks must have their parent block as first element in the parent list.

Hum, I'm still a little concerned because the DAG can get really deep sometimes. Maybe add a protection that will fail and report a problem or move to a DFS implementation using a stack (instead of a recursion).

Reply:
Created an issue for this refactor here at https://github.com/HathorNetwork/hathor-wallet-service-sync_daemon/issues/6